src/database_lmdb.cpp

```diff
@@ -0,0 +1,341 @@
+#include <errno.h>
+
+#define MAKEKEY(pos)                    \
+	MDB_val key;                        \
+	s64 key_ = getBlockAsInteger(pos);  \
+	key.mv_size = sizeof(s64);          \
+	key.mv_data = &key_;
+
+//public:
+Database_LMDB::Database_LMDB(const std::string &savedir, Settings &conf)
+{
+	int result;
+
+	online = false;
+	path = std::string (savedir) + DIR_DELIM + "map.mdb";
+	configuration = &conf;
+
+	mapSize = conf.exists("lmdb_mapsize") ? conf.getU64("lmdb_mapsize") :
+		defaultMapSize;
+
+	if ((result = mdb_env_create(&environment))) {
+		except("LMDB: Unknown error occurred creating environment.");
+	}
+
+	if ((result = mdb_env_set_mapsize(environment, mapSize))) {
+		// Most probably, this means the mapsize is invalid.
+		if (mdb_env_set_mapsize(environment, defaultMapSize)) {
+			closeDatabase();
+			except("LMDB: Can't set mapsize.");
+		}
+	}
+
+	// Flag justification:  NOMETASYNC allows a commit to occur with only
+	// one flush.  The price is durability -- the most recently committed
+	// transaction may be lost in the event of an error that prevents the
+	// buffers from making it to disk.  Probably not a problem for us.
+	// In no case will this result in a corrupted database, since LMDB
+	// never overwrites live data.
+	if ((result = mdb_env_open(environment, path.c_str(),
+			MDB_NOMETASYNC|MDB_NOSUBDIR, 0600))) {
+		closeDatabase();
+		if ((result == MDB_VERSION_MISMATCH)) {
+			except("LMDB: Library version doesn't match the one that "
+				"created the environment.");
+		}
+		else if (result == MDB_INVALID) {
+			except("LMDB: Environment headers corrupted.");
+		}
+		else if (result == ENOENT) {
+			except("LMDB: World directory removed before LMDB could open "
+				"database.");
+		}
+		else if (result == EACCES) {
+			except("LMDB: Inappropriate permissions on LMDB database.");
+		}
+		else if (result == EAGAIN) {
+			except("LMDB: Database locked by another process.");
+		}
+		else {
+			except("LMDB: Unknown error in env_open.");
+		}
+	}
+
+	makeTransaction(MDB_RDONLY);
+
+	if ((result = mdb_dbi_open(transaction, NULL, MDB_INTEGERKEY,
+			&database))) {
+		closeDatabase();
+		except("LMDB: Can't open initial database.");
+	}
+	online = true;
+}
+
+
+Database_LMDB::~Database_LMDB()
+{
+	closeDatabase();
+}
+
+
+void Database_LMDB::beginSave()
+{
+	mdb_txn_abort(transaction);
+	makeTransaction(0);
+}
+
+
+void Database_LMDB::endSave()
+{
+	commit();
+	makeTransaction(MDB_RDONLY);
+}
+
+
+bool Database_LMDB::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	int result;
+	MDB_val value;
+	size_t temp_size = data.length();
+	std::string temp_data = data;
+
+	MAKEKEY(pos);
+
+	value.mv_data = const_cast<char *>(temp_data.c_str());
+	value.mv_size = temp_size;
+
+	while ((result = mdb_put(transaction, database, &key, &value, 0))) {
+		if (result == MDB_TXN_FULL) {
+			commit();
+			makeTransaction(0);
+			continue;
+		}
+		else if (result == MDB_MAP_FULL) {
+			commit();
+			mapSize += pow(1024,2);
```

Comment:
I don't think it would lead to more fragmentation.

I suspect that LMDB maps the database in memory contiguously, so fragmentation there would not be an issue (if I'm correct). And as the file on disk grows incrementally anyway, it will probably be just as fragmented.

As growing is not normally needed on Windows either, provided that LMDB version 0.9.18 or later is used (see another comment above), [this](http://www.openldap.org/lists/openldap-technical/201303/msg00145.html) message (also referenced above), suggests that it may be better to set the default size to huge. 1TB for instance on 64 bit.